server/src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/random/RandomSamplerAggregatorFactory.java

```diff
@@ -87,11 +87,10 @@ private Weight getWeight() throws IOException {
                 shardSeed == null ? context.shardRandomSeed() : shardSeed
             );
             BooleanQuery booleanQuery = new BooleanQuery.Builder().add(query, BooleanClause.Occur.FILTER)
-                .add(context.query(), BooleanClause.Occur.FILTER)
+                .add(context.query(), BooleanClause.Occur.MUST)
```

Comment:
These changes propagate the scoring function from the original query through the random sampler to the subaggs.

I'm not sure this is maximally efficient: if the subaggs don't need scores, this might trigger doing unnecessary work.

Reply:
I think doing a `MUST` always is OK. We will rewrite it appropriately as necessary back to a filter if it doesn't require scoring.

As for `COMPLETE` vs `COMPLETE_NO_SCORES` we should update this weight factory supplier to accept the sub-aggregators. The sub-aggregators shouldn't change throughout the life-time of the aggregator on the shard. So continuing to have the same weight once its cached is ok. 

But, I would take a look at the sub-aggregators, and their provided `scoreMode()` values. This will indicate to us if we need to provide `MUST` & `COMPLETE` vs `FILTER` & `COMPLETE_NO_SCORES`.

There are examples of this all around the code base,

```
        ScoreMode scoreMode = null;
        for (BucketCollector collector : collectors) {
            if (scoreMode == null) {
                scoreMode = collector.scoreMode();
            } else if (scoreMode != collector.scoreMode()) {
                return ScoreMode.COMPLETE;
            }
        }
        return scoreMode;
```
^ how the multi-bucket collectors work. 

But aggregators will provide `scoreMode()` and the score mode we use should be based off that if possible

Reply:
Thanks. I'm glad I seem to be on the right track.

I had looked into something like you suggest. However, regarding the sub-aggregators, in this `RandomSamplerAggregatorFactory` I only have access to `AggregatorFactories.Builder subFactories`. They don't expose the `ScoreMode`; that's a method in the `AggregatorBase`. Any idea how to access this without a major overhaul of the code?

Reply:
The best thing I see right now is adding a method `scoreMode()` to the `AggregationBuilder`. What do you think of that, @benwtrent ?